core/templates/components/skill-selector/skill-selector.component.ts

```diff
@@ -0,0 +1,213 @@
+     { subTopicName: string; checked: boolean }[] } = {};
+  intialSubTopicFilterDict = {};
+
+  private filterForMatchingSubtringPipe: FilterForMatchingSubstringPipe = (
+    new FilterForMatchingSubstringPipe()
+  );
+
+  constructor() {}
+
+  ngOnInit(): void {
+    this.currCategorizedSkills = this.categorizedSkills;
+    for (let topicName in this.currCategorizedSkills) {
+      var topicNameDict = {
+        topicName: topicName,
+        checked: false
+      };
+      this.topicFilterList.push(topicNameDict);
+      var subTopics = this.currCategorizedSkills[topicName];
+      this.subTopicFilterDict[topicName] = [];
+      for (let subTopic in subTopics) {
+        var subTopicNameDict = {
+          subTopicName: subTopic,
+          checked: false
+        };
+        this.subTopicFilterDict[topicName].push(subTopicNameDict);
+      }
+    }
+    this.intialSubTopicFilterDict = angular.copy(this.subTopicFilterDict);
+  }
+
+  checkIfEmpty(skills: string[]): boolean {
+    return skills.length === 0;
+  }
+
+  checkTopicIsNotEmpty(topicName: string): boolean {
+    for (let key in this.currCategorizedSkills[topicName]) {
+      if (Object.keys(this.currCategorizedSkills[topicName][key]).length) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  setSelectedSkillId(): void {
+    this.selectedSkillIdChanged.emit(this.selectedSkill);
+  }
+
+  // The folowing function is called when the subtopic filter changes.
+  // This updates the list of Skills displayed in the selector.
+  updateSkillsListOnSubtopicFilterChange(): void {
+    var updatedSkillsDict = {};
+    var isAnySubTopicChecked = false;
+    for (let topicName in this.subTopicFilterDict) {
+      var subTopics = this.subTopicFilterDict[topicName];
+      for (var i = 0; i < subTopics.length; i++) {
+        if (subTopics[i].checked) {
+          if (!updatedSkillsDict.hasOwnProperty(topicName)) {
+            updatedSkillsDict[topicName] = {};
+          }
+          var tempCategorizedSkills = this.categorizedSkills;
+          var subTopicName = subTopics[i].subTopicName;
+          updatedSkillsDict[topicName][subTopicName] =
+            tempCategorizedSkills[topicName][subTopicName];
+          isAnySubTopicChecked = true;
+        }
+      }
+    }
+    if (!isAnySubTopicChecked) {
+      // If no subtopics are checked in the subtop filter, we have
+      // to display all the skills from checked topics.
+      var isAnyTopicChecked = false;
+      for (var i = 0; i < this.topicFilterList.length; i++) {
+        if (this.topicFilterList[i].checked) {
+          var tempCategorizedSkills = this.categorizedSkills;
+          var topicName: string = this.topicFilterList[i].topicName;
+          updatedSkillsDict[topicName] = tempCategorizedSkills[topicName];
+          isAnyTopicChecked = true;
+        }
+      }
+      if (isAnyTopicChecked) {
+        this.currCategorizedSkills = angular.copy(updatedSkillsDict);
```

Comment:
angular.copy() is only available in angularjs.

Reply:
Done